src/pip/_internal/cli/main_parser.py

```diff
@@ -45,6 +49,44 @@ def create_main_parser() -> ConfigOptionParser:
     return parser
 
 
+def identify_python_interpreter(python: str) -> Optional[str]:
+    if python == "python" or python == "py":
+        # Run the active Python.
+        # We have to be very careful here, because:
+        #
+        # 1. On Unix, "python" is probably correct but there is a "py" launcher.
+        # 2. On Windows, "py" is the best option if it's present.
+        # 3. On Windows without "py", "python" might work, but it might also
+        #    be the shim that launches the Windows store to allow you to install
+        #    Python.
+        #
+        # We go with getting py on Windows, and if it's not present or we're
+        # on Unix, get python. We don't worry about the launcher on Unix or
+        # the installer stub on Windows.
+        py = None
+        if WINDOWS:
+            py = shutil.which("py")
+        if py is None:
+            py = shutil.which("python")
+        if py:
+            return py
+
+    # TODO: On Windows, `--python .venv/Scripts/python` won't pass the
+    #       exists() check (no .exe extension supplied). But it's pretty
+    #       obvious what the user intends. Should we allow this?
+    if os.path.exists(python):
+        if not os.path.isdir(python):
+            return python
```

Comment:
Should we check that it exists here ?
Should we check if it has the executable bit, or is the error printed by the subprocess call sufficiently clear ?

Reply:
If we switch this to shutil.which, as proposed above, we don’t have to deal with that nuance. :)

Reply:
> Should we check that it exists here ?

Line 77 confirms it exists, line 78 is simply checking it's not a directory (because we look a bit further if it is)...

Executable bit doesn't exist on Windows, so I never thought about it. Is the generic message "Could not find python interpreter {python}" good enough when the file exists but is not executable? I'm somewhat reluctant to complicate things with a custom message, but I'm OK with simply ckecking `os.access(python, os.X_OK)` and returning `None` if that's false. Everything seems to test as `X_OK` on Windows, so I guess this is OK (nothing will be rejected incorrectly).

Reply:
It is probably simpler to catch exceptions from `subprocess.run`. The error message we get in the exception should be clear enough but we probably want to avoid the stack trace if the file is not executable or any other reason for not being able to run it.

Reply:
OK. I've just added the `os.access` check but I'll follow up with something like you suggest. I'll do a little experimenting to see what sort of errors `subprocess.run` can return.

Reply:
> The error message we get in the exception should be clear enough

Hmm, if I pass something that isn't an executable file (`setup.cfg`) to `subprocess.run`, Windows returns the delightful error

```
OSError: [WinError 193] %1 is not a valid Win32 application
```

And yes, that `%1` placeholder is returned literally, it's *not* filled in with the filename :-(

On Unix, I get

```
PermissionError: [Errno 13] Permission denied: './t.py'
```

What's weird on Unix is that if I specify a filename in the current directory *without* `./`, I get

```
FileNotFoundError: [Errno 2] No such file or directory: 't.py'
```

whether or not the file has execute permission. I assume that's because it treats a path with no directory component as needing a `PATH` search, and the current directory isn't on `PATH`. So I think I need to do an `abspath()` on the path the user supplied.

I think the reasonable thing to do here is to just bite the bullet and do extra work to give a good error:

1. Make the supplied interpreter name absolute using `abspath()`.
2. Retain the executable check, but raise an explicit error "Supplied interpreter name exists but is not executable".
3. Catch errors from `subprocess.run` and report them as `Could not run Python interpreter {name}: {exception}`.

Reply:
>  think the reasonable thing to do here is to just bite the bullet 

Or maybe not? As long as there is no scary stack trace, it may not be necessary to try to be smarter than the OS, as these errors should be rare anyway.